flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/materializedtable/MaterializedTableManager.java

```diff
@@ -1244,4 +1278,99 @@ private static List<RowData> fetchAllResults(ResultFetcher resultFetcher) {
         }
         return results;
     }
+
+    private JobExecutionResult executeJob(
+            String script,
+            Configuration executionConfig,
+            OperationExecutor operationExecutor,
+            OperationHandle operationHandle) {
+        if (isApplicationMode(operationExecutor.getSessionContext().getSessionConf())) {
+            return executeApplicationJob(script, executionConfig, operationExecutor);
+        } else {
+            return executeSessionJob(script, executionConfig, operationExecutor, operationHandle);
+        }
+    }
+
+    private boolean isApplicationMode(Configuration configuration) {
+        String target = configuration.get(TARGET);
+        return target.endsWith("application");
+    }
+
+    private JobExecutionResult executeSessionJob(
+            String script,
+            Configuration executionConfig,
+            OperationExecutor operationExecutor,
+            OperationHandle operationHandle) {
+        Configuration mergedConfig =
+                new Configuration(operationExecutor.getSessionContext().getSessionConf());
+        mergedConfig.addAll(executionConfig);
+
+        // submit flink streaming job
+        ResultFetcher resultFetcher =
+                operationExecutor.executeStatement(operationHandle, mergedConfig, script);
+
+        // get execution.target and jobId, clusterId
+        List<RowData> results = fetchAllResults(resultFetcher);
+        String jobId = results.get(0).getString(0).toString();
+        String executeTarget = operationExecutor.getSessionContext().getSessionConf().get(TARGET);
+        Optional<String> clusterId = operationExecutor.getSessionClusterId();
+
+        return new JobExecutionResult(jobId, executeTarget, clusterId.orElse(null));
+    }
+
+    private JobExecutionResult executeApplicationJob(
+            String script, Configuration executionConfig, OperationExecutor operationExecutor) {
+        List<String> arguments = new ArrayList<>();
+        arguments.add("--" + SqlDriver.OPTION_SQL_SCRIPT.getLongOpt());
+        arguments.add(script);
+
+        Configuration mergedConfig =
+                new Configuration(operationExecutor.getSessionContext().getSessionConf());
+        mergedConfig.addAll(executionConfig);
+        JobID jobId = new JobID();
+        mergedConfig.set(PIPELINE_FIXED_JOB_ID, jobId.toString());
```

Comment:
We must set jobId in advance for application mode?

Reply:
If not set, jobId will be generated after being submitted to the application cluster. On the one hand, this is an asynchronous logic. At the same time, there is no good way to obtain it. Therefore, we need to generate jobId in advance.